### ISSUE TITLE
We need to get a package with information about the tile.

### ISSUE THREAD

**Issue by lamo2k123** [kevinhikaruevans/uojs2#45]

**Comment by kevinhikaruevans** [comment]
I'm working on this now. :)

Right now, I have a script that integrates with servuo and it returns map data, tiledata, and art.

**Comment by kevinhikaruevans** [comment]
@lamo2k123 do you think it is better to get JSON map data & tile data with REST API or through websockets?

**Comment by lamo2k123** [comment]
Websocket

**Comment by lamo2k123** [comment]
tile texture gives the server? or server gives texture id?

**Comment by lamo2k123** [comment]
png gives the game server?

**Comment by kevinhikaruevans** [comment]
Yes... For now.


The image/data server runs through ServUO. It can run independently too though.

**Comment by lamo2k123** [comment]
I thought that all the static is stored in the client.

**Comment by kevinhikaruevans** [comment]
It is, but it would make more sense to have a server that can serve images for the game client.

**Comment by lamo2k123** [comment]
Hmmm....

**Comment by kevinhikaruevans** [comment]
Static content is over 1GB, so the client would need to download it every load. 

The server can send images as needed... so many smaller requests vs one big request...

**Comment by lamo2k123** [comment]
how much more work on this task?

**Comment by kevinhikaruevans** [comment]
I'll try getting something up tonight after work :)

**Comment by kevinhikaruevans** [comment]
Here's an example of the map data:

http://107.161.24.129:2590/map?y=500&x=500

I can't get images working. It looks like Mono/libgdiplus doesn't support ARGB1555, which is needed for Ultima.dll to get images.

**Comment by lamo2k123** [comment]
😞

**Comment by kevinhikaruevans** [comment]
I'm updating Ultima.dll to use ARGB8888 instead of ARGB1555 later tonight, so we should be able to get images working.

**Comment by kevinhikaruevans** [comment]
Loads with ARGB8888 👍 

http://107.161.24.129:2590/land?id=4
http://107.161.24.129:2590/land?id=168

**Comment by kevinhikaruevans** [comment]
@lamo2k123 - should I wait for you to finish the refactoring before starting implementing the graphics?

**Comment by lamo2k123** [comment]
Mmm... Use branch 'refactoring'.

**Comment by lamo2k123** [comment]
Hmm.. Origin client size textures 64x64 and 128x128. And rotate 0

**Comment by kevinhikaruevans** [comment]
In 'refactoring', I npm installed `history`, `hook-redux`, and `react-router-async`. 

Any other packages needed?

In console:

> [HMR] Waiting for update signal from WDS...
> index.js:111 Object {length: 2, action: "POP", location: Object}
> core.js:97  action @ 20:12:34.197 @@router/START_TRANSITION (in 0.20 ms)
> index.js:291 Uncaught (in promise) TypeError: route.action is not a function
>     at Router.<anonymous> (index.js:291)
>     at step (index.js:40)
>     at Object.next (index.js:21)
>     at fulfilled (index.js:12)
> (anonymous) @ index.js:291
> step @ index.js:40
> (anonymous) @ index.js:21
> fulfilled @ index.js:12
> client?8505:38 [WDS] Hot Module Replacement enabled.
> client?8505:38 [WDS] Warnings while compiling.

**Comment by kevinhikaruevans** [comment]
Sweet, thank you. :)

**Comment by kevinhikaruevans** [comment]
Have you used https://www.npmjs.com/package/request before?

I need to grab JSON and wondering if this is good to install or not. Could also just use XMLHttpRequest too.

**Comment by lamo2k123** [comment]
Use the 'window.fetch'

**Comment by lamo2k123** [comment]
which package contains data on the location of the character on the map?

**Comment by kevinhikaruevans** [comment]
I think it is either packet 0x1B or packet 0x20:

https://github.com/kevinhikaruevans/uojs2/blob/master/src/client/state/player/actions.js#L34

https://github.com/kevinhikaruevans/uojs2/blob/master/src/client/state/player/actions.js#L54

**Comment by lamo2k123** [comment]
http://107.161.24.129:2590/map?y=3503&x=2574 here are the coordinates I got

**Comment by lamo2k123** [comment]
there mapId need?

**Comment by kevinhikaruevans** [comment]
it's using blocks, so you'll need to do 3503/8 = 437, 2574/8 = 321, so you'll get:

http://107.161.24.129:2590/map?y=437&x=321

There is a map ID but I need to implement it still. Right now, it's using Felucca.

**Comment by kevinhikaruevans** [comment]
We'll probably need to create a system to detect as the player moves, preload each map block when they move near it.

**Comment by lamo2k123** [comment]
![screen](https://cloud.githubusercontent.com/assets/674847/21532204/2dfa08c4-cd5f-11e6-8068-d673041ca3ed.JPG)
![playerg_12-29_00 06](https://cloud.githubusercontent.com/assets/674847/21532206/314e4c92-cd5f-11e6-813f-502549d226f4.jpg)
8x8 block small :(((

**Comment by kevinhikaruevans** [comment]
Yeah, block is too small, so we'll have to load multiple blocks at once. 

I'm trying to import THREE: https://github.com/kevinhikaruevans/uojs2/compare/refactoring...refactoring-map-additions

but `import * as THREE from 'three'` doesn't seem to work. Any ideas? 😕

**Comment by TorchwoodPL** [comment]
Do you commit the res server for me to hook up on my vps? :)

**Comment by kevinhikaruevans** [comment]
In ServUO's Scripts directory:

[UOJSServer.zip](https://github.com/kevinhikaruevans/uojs2/files/676650/UOJSServer.zip)

Then in Data/Assemblies.cfg, I'm including these:

```
Newtonsoft.Json.dll
websocket-sharp.dll
```

(you'll need to download the binary dlls)

**Comment by lamo2k123** [comment]
And where you can view the source code for parsing files map.mul

**Comment by kevinhikaruevans** [comment]
I'm using Ultima.dll to do that.

**Comment by lamo2k123** [comment]
can nodejs to try to parse map file and make the api?

**Comment by kevinhikaruevans** [comment]
Yeah, but it'll take a long time to do.

**Comment by lamo2k123** [comment]
I would like to just getting the tiles to rewrite.

**Comment by kevinhikaruevans** [comment]
It's just a matter of porting over Ultima.dll to nodejs. Should I work on doing that?

https://github.com/ServUO/ServUO/tree/master/Ultima

**Comment by lamo2k123** [comment]
I want more convenient and flexible API to obtain data on the surrounding tiles. But we need to understand how to read the map file map (x) .uop

**Comment by kevinhikaruevans** [comment]
It's being done here: https://github.com/ServUO/ServUO/blob/master/Ultima/TileMatrix.cs#L414


Should I create repo for nodejs port package?

**Comment by lamo2k123** [comment]
created. I think it will take much time.

**Comment by kevinhikaruevans** [comment]
Do you have a link for it?

**Comment by lamo2k123** [comment]
I did not create. you create the repository.?

**Comment by kevinhikaruevans** [comment]
Ooh, just did:

https://github.com/kevinhikaruevans/uodatareader

**Comment by lamo2k123** [comment]
nice 👍

**Comment by kevinhikaruevans** [comment]
Should it use uodatareader in server/index.js and send map responses through the websocket? Then we can send map JSON data and stuff through it.

**Comment by lamo2k123** [comment]
of course yes

**Comment by kevinhikaruevans** [comment]
What about things like the images? Do we need to setup a separate webserver (like Express) to host this?

**Comment by lamo2k123** [comment]
hmm.. I do not know yet. I would like to work in the production is not static nodejs distributed. I have not had time to look 'uodatareader' :) and I can not yet imagine the interaction of components.

**Comment by lamo2k123** [comment]
how do you plan to implement threejs Z Index tiles?

**Comment by kevinhikaruevans** [comment]
I'm just setting the z in the position on the tile. We get the map data `[{id: 123, z: 10}, ...]`, then use that z in the coordinate (`tileposition = new THREE.Vector3(someX, someY, tile.z)`).

It gets set here:

https://github.com/kevinhikaruevans/uojs2/blob/0d62a44543d5d0b21a5422f62aea5b504414a714/src/client/components/map/index.js#L34

I need to figure out the camera rotation for it to look better here: 

https://github.com/kevinhikaruevans/uojs2/blob/0d62a44543d5d0b21a5422f62aea5b504414a714/src/client/components/scene/index.js#L30

It's rotating the camera along the Z to 45-degrees, but then it's rotating it along the X axis to give the Z axis depth.

**Comment by kevinhikaruevans** [comment]
Right now, it's looking like this for me:

![image](https://cloud.githubusercontent.com/assets/4730591/22312019/0d594668-e30b-11e6-8f85-026861380411.png)
